### PR TITLE
Creating Dimension class that is required for multi-dim Measurements

### PR DESCRIPTION
**Pull request by kdsudac**
This change is [Reviewable](https://reviewable.io/reviews/google/openhtf/710)

### CODE REVIEW

File: openhtf/core/measurements.py [hunk 1]
```diff
@@ -304,6 +317,57 @@ def set(self, value):
...
+  """Dimension for multi-dim Measurements.
+
+  Dimensions optionally include a unit and a description.  This is intended
+  as a drop-in replacement for UnitDescriptor for backwards compatibility.
+  """
+
+  def __init__(self, description='', unit=units.NO_DIMENSION):
+    self.description = description
+    self.unit = unit
+
```
Comment (fahhem) [inline openhtf/core/measurements.py]: extra line

**Comment by fahhem** [comment]
I don't have context on this change's reason to exist, so I'll let someone with that context merge it in.

File: openhtf/__init__.py [hunk 2]
```diff
@@ -39,7 +39,7 @@
 from openhtf import core
 from openhtf import plugs
 from openhtf import util
-from openhtf.core.measurements import Measurement, measures
+from openhtf.core.measurements import Measurement, measures, Dimension
```
Comment (Kenadia) [inline openhtf/__init__.py]: import order `Dimension, Measurement, measures`

File: openhtf/core/measurements.py [hunk 3]
```diff
@@ -304,6 +317,56 @@ def set(self, value):
...
+  def _asdict(self):
+    return {
+      'code': self.code,
+      'description': self.description,
+      'name': self.name,
+      'suffix': self.suffix,
+    }
+
+
+
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: extra line

File: openhtf/core/measurements.py [hunk 4]
```diff
@@ -304,6 +317,56 @@ def set(self, value):
...
+  def name(self):
+    """Provides backwards compatibility to `units.UnitDescriptor` api."""
+    return self.description or self.unit.name
+
+  def _asdict(self):
+    return {
+      'code': self.code,
+      'description': self.description,
+      'name': self.name,
+      'suffix': self.suffix,
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: +2 indent

File: openhtf/core/measurements.py [hunk 5]
```diff
@@ -304,6 +317,56 @@ def set(self, value):
...
+    self.unit = unit
+
+  @classmethod
+  def from_unit_descriptor(cls, unit_desc):
+    return cls(unit=unit_desc)
+
+  @classmethod
+  def from_string(cls, string):
+    """Convert a string into a Dimension"""
+    # Note:There is some ambiguity as to whether the string passed is intended
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: `Note: ...`

File: openhtf/core/measurements.py [hunk 6]
```diff
@@ -185,6 +185,19 @@ def _maybe_make_unit_desc(self, unit_desc):
                                                                 unit_desc))
     return unit_desc
 
+  def _maybe_make_dimension(self, dimension):
+    """Return a `measurements.Dimension` instance."""
+    # For backwards compatibility the dimension can be either a Dimension, a
+    # string or a `units.UnitDescriptor`.
```
Comment (Kenadia) [inline openhtf/core/measurements.py]: To me this reads more clearly as:

```
For backwards compatibility the argument can be either...
```

File: openhtf/__init__.py (hunk 2 above)
Comment (kdsudac) [reply openhtf/__init__.py]: Done.

File: openhtf/core/measurements.py (hunk 6 above)
Comment (kdsudac) [reply openhtf/core/measurements.py]: Done.

File: openhtf/core/measurements.py (hunk 5 above)
Comment (kdsudac) [reply openhtf/core/measurements.py]: Done.

File: openhtf/core/measurements.py (hunk 4 above)
Comment (kdsudac) [reply openhtf/core/measurements.py]: Done.

File: openhtf/core/measurements.py (hunk 3 above)
Comment (kdsudac) [reply openhtf/core/measurements.py]: Done.

File: openhtf/core/measurements.py (hunk 1 above)
Comment (kdsudac) [reply openhtf/core/measurements.py]: Done.

**Comment by grybmadsci** [comment]
For a little context here, the request was to add custom units to differentiate columns with the same units in a multidim measurement.  The pushback was from me that I prefer to keep the units strongly standardized, and proposed adding a string description to the dimension metadata as an alternative.

**Pull request merged**